memecity.engine.ecs/EntityManager.h

```diff
@@ -1,54 +1,137 @@
-	vector<Entity*> get_entities_with_component(string type);
+		///<summary>Get all entities with a certain component.</summary>
+		std::vector<std::reference_wrapper<Entity>> get_entities_with_component(component_typetoken token)
+		{
+			std::vector<std::reference_wrapper<Entity>> result;
+			for (auto& e : entities)
+				if (this->has_component(e, token))
+					result.push_back(std::ref(e));
+			return result;
+		}
 
-	///<summary>Get components of a specific type.</summary>
-	vector<Component*> get_components_of_type(string type);
+		///<summary>Get components of a specific type.</summary>
+		template<class C>
+		std::vector<std::reference_wrapper<C>> get_components_of_type(component_typetoken token)
+		{
+			static_assert(std::is_convertible<C*, Component*>::value, "This function can only retrieve concrete subclasses of Component");
 
-	///<summary>Get specific component of entity</summary>
-	Component* get_component_of_entity(int entity_id, string type);
+			std::vector<std::reference_wrapper<C>> result;
+			for (auto& c : components[token])
+				result.push_back(std::ref(*(dynamic_cast<C*>(c.get()))));
+			return result;
+		}
 
-	///<summary>Get components with a specific entity ID.</summary>
-	vector<Component*> get_components_of_entity(int entity_id);
+		///<summary>Get specific component of entity</summary>
+		template<class C>
+		C* get_component_of_entity(Entity& entity, component_typetoken token)
+		{
+			static_assert(std::is_convertible<C*, Component*>::value, "This function can only retrieve concrete subclasses of Component");
+			for (auto& pairs : components)
+				for (auto& c : pairs.second)
+					if (c->entity == entity && c->get_type_token() == token)
+						return dynamic_cast<C*>(c.get());
+			return nullptr;
+		}
 
-	///<summary>Checks if entity has component.</summary>
-	bool has_component(Entity* e, string type);
+		///<summary>Get components with a specific entity ID.</summary>
+		template<class C>
+		std::vector<std::reference_wrapper<C>> get_components_of_entity(Entity& entity)
+		{
+			static_assert(std::is_convertible<C*, Component*>::value, "This function can only retrieve concrete subclasses of Component");
+			std::vector<std::reference_wrapper<C>> result;
+			for (auto& pairs : components)
+				for (auto& c : pairs.second)
+					if (c->entity == entity)
+						result.push_back(std::ref(*(dynamic_cast<C*>(c.get()))));
+			return result;
+		}
 
-	///<summary>Run all systems.</summary>
-	void update();
+		///<summary>Checks if entity has component.</summary>
+		bool has_component(Entity& entity, component_typetoken token)
+		{
+			for (auto& c : components[token])
+				if (c->entity == entity)
+					return true;
+			return false;
+		}
+
+		///<summary>Run all systems.</summary>
+		void update()
+		{
+			for (auto& pair : systems)
+				pair.second->run(*this);
+		}
+
+		~EntityManager()
+		{
+			entities.clear();
+			for (auto& pair : components)
+			{
+				pair.second.clear();
+			}
```

Comment:
This for each loop has curly braces while the other ones don't. Please be consistent.